### ISSUE TITLE
Doesn't support scala3?

### ISSUE THREAD

**Issue by qwqawawow** [konsoletyper/teavm#647]
```
[ERROR] Method java.lang.invoke.MethodHandle.invoke()V was not found
    at scala.runtime.Statics.releaseFence(Statics.java:148)
    at scala.collection.immutable.$colon$colon.<init>(List.scala:656)
    at scala.collection.immutable.List.$colon$colon(List.scala:97)
    at scala.util.CommandLineParser$.parseRemainingArguments(CommandLineParser.scala:33)                                                                  at Main.main(Main.scala:1)
[ERROR] Method java.lang.invoke.MethodHandles.lookup()Ljava/lang/invoke/MethodHandles$Lookup; was not found
    at scala.runtime.Statics$VM.mkHandle(Statics.java:159)                     at scala.runtime.Statics$VM.<clinit>(Statics.java:155)
    at scala.runtime.Statics.releaseFence(Statics.java:148)
    at scala.collection.immutable.$colon$colon.<init>(List.scala:656)
    at scala.collection.immutable.List.$colon$colon(List.scala:97)
    at scala.util.CommandLineParser$.parseRemainingArguments(CommandLineParser.scala:33)
    at Main.main(Main.scala:1)
[ERROR] Class java.lang.ExceptionInInitializerError was not found              at scala.runtime.Statics$VM.mkHandle(Statics.java:172)
    at scala.runtime.Statics$VM.<clinit>(Statics.java:155)                     at scala.runtime.Statics.releaseFence(Statics.java:148)
    at scala.collection.immutable.$colon$colon.<init>(List.scala:656)          at scala.collection.immutable.List.$colon$colon(List.scala:97)
    at scala.util.CommandLineParser$.parseRemainingArguments(CommandLineParser.scala:33)
    at Main.main(Main.scala:1)
[ERROR] Class java.lang.ExceptionInInitializerError was not found
    at scala.runtime.Statics$VM.mkHandle(Statics.java:167)
    at scala.runtime.Statics$VM.<clinit>(Statics.java:155)                     at scala.runtime.Statics.releaseFence(Statics.java:148)
    at scala.collection.immutable.$colon$colon.<init>(List.scala:656)
    at scala.collection.immutable.List.$colon$colon(List.scala:97)
    at scala.util.CommandLineParser$.parseRemainingArguments(CommandLineParser.scala:33)
    at Main.main(Main.scala:1)                                             [ERROR] Method java.lang.invoke.MethodType.methodType(Ljava/lang/Class;)Ljava/lang/invoke/MethodType; was not found
    at scala.runtime.Statics$VM.mkHandle(Statics.java:161)                     at scala.runtime.Statics$VM.<clinit>(Statics.java:155)
    at scala.runtime.Statics.releaseFence(Statics.java:148)                    at scala.collection.immutable.$colon$colon.<init>(List.scala:656)
    at scala.collection.immutable.List.$colon$colon(List.scala:97)
    at scala.util.CommandLineParser$.parseRemainingArguments(CommandLineParser.scala:33)                                                                  at Main.main(Main.scala:1)
[ERROR] Method java.lang.invoke.MethodType.methodType(Ljava/lang/Class;)Ljava/lang/invoke/MethodType; was not found
    at scala.runtime.Statics$VM.mkHandle(Statics.java:165)
    at scala.runtime.Statics$VM.<clinit>(Statics.java:155)                     at scala.runtime.Statics.releaseFence(Statics.java:148)
    at scala.collection.immutable.$colon$colon.<init>(List.scala:656)          at scala.collection.immutable.List.$colon$colon(List.scala:97)
    at scala.util.CommandLineParser$.parseRemainingArguments(CommandLineParser.scala:33)                                                                  at Main.main(Main.scala:1)
[ERROR] Method java.util.Set.copyOf(Ljava/util/Collection;)Ljava/util/Set; was not found
    at jdk.internal.reflect.Reflection.registerFilter(Reflection.java:291)
    at jdk.internal.reflect.Reflection.registerMethodsToFilter(Reflection.java:280)                                                                       at sun.misc.Unsafe.<clinit>(Unsafe.java:59)
    at java.lang.Class.initialize
    at scala.runtime.Statics$VM.mkHandle(Statics.java:161)
    at scala.runtime.Statics$VM.<clinit>(Statics.java:155)
    at scala.runtime.Statics.releaseFence(Statics.java:148)
    at scala.collection.immutable.$colon$colon.<init>(List.scala:656)
    at scala.collection.immutable.List.$colon$colon(List.scala:97)
    at scala.util.CommandLineParser$.parseRemainingArguments(CommandLineParser.scala:33)
    at Main.main(Main.scala:1)
[ERROR] Class java.lang.Module was not found
    at jdk.internal.reflect.Reflection.<clinit>(Reflection.java:55)
    at sun.misc.Unsafe.<clinit>(Unsafe.java:59)                                at java.lang.Class.initialize
    at scala.runtime.Statics$VM.mkHandle(Statics.java:161)
    at scala.runtime.Statics$VM.<clinit>(Statics.java:155)
    at scala.runtime.Statics.releaseFence(Statics.java:148)                    at scala.collection.immutable.$colon$colon.<init>(List.scala:656)
    at scala.collection.immutable.List.$colon$colon(List.scala:97)
    at scala.util.CommandLineParser$.parseRemainingArguments(CommandLineParser.scala:33)
    at Main.main(Main.scala:1)                                             [ERROR] Native method jdk.internal.misc.Unsafe.registerNatives()V has no implementation
    at jdk.internal.misc.Unsafe.registerNatives
    at jdk.internal.misc.Unsafe.<clinit>(Unsafe.java:60)
    at sun.misc.Unsafe.<clinit>(Unsafe.java:65)
    at java.lang.Class.initialize
    at scala.runtime.Statics$VM.mkHandle(Statics.java:161)
    at scala.runtime.Statics$VM.<clinit>(Statics.java:155)
    at scala.runtime.Statics.releaseFence(Statics.java:148)
    at scala.collection.immutable.$colon$colon.<init>(List.scala:656)
    at scala.collection.immutable.List.$colon$colon(List.scala:97)             at scala.util.CommandLineParser$.parseRemainingArguments(CommandLineParser.scala:33)
    at Main.main(Main.scala:1)                                             [ERROR] Native method jdk.internal.misc.Unsafe.arrayBaseOffset0(Ljava/lang/Class;)I has no implementation
    at jdk.internal.misc.Unsafe.arrayBaseOffset0
    at jdk.internal.misc.Unsafe.arrayBaseOffset(Unsafe.java:1173)
    at jdk.internal.misc.Unsafe.<clinit>(Unsafe.java:1179)                     at sun.misc.Unsafe.<clinit>(Unsafe.java:65)
    at java.lang.Class.initialize                                              at scala.runtime.Statics$VM.mkHandle(Statics.java:161)
    at scala.runtime.Statics$VM.<clinit>(Statics.java:155)
    at scala.runtime.Statics.releaseFence(Statics.java:148)                    at scala.collection.immutable.$colon$colon.<init>(List.scala:656)
    at scala.collection.immutable.List.$colon$colon(List.scala:97)
    at scala.util.CommandLineParser$.parseRemainingArguments(CommandLineParser.scala:33)
    at Main.main(Main.scala:1)                                             [ERROR] Native method jdk.internal.misc.Unsafe.arrayIndexScale0(Ljava/lang/Class;)I has no implementation
    at jdk.internal.misc.Unsafe.arrayIndexScale0                               at jdk.internal.misc.Unsafe.arrayIndexScale(Unsafe.java:1229)
    at jdk.internal.misc.Unsafe.<clinit>(Unsafe.java:1235)
    at sun.misc.Unsafe.<clinit>(Unsafe.java:65)
    at java.lang.Class.initialize
    at scala.runtime.Statics$VM.mkHandle(Statics.java:161)                     at scala.runtime.Statics$VM.<clinit>(Statics.java:155)
    at scala.runtime.Statics.releaseFence(Statics.java:148)
    at scala.collection.immutable.$colon$colon.<init>(List.scala:656)          at scala.collection.immutable.List.$colon$colon(List.scala:97)
    at scala.util.CommandLineParser$.parseRemainingArguments(CommandLineParser.scala:33)                                                                  at Main.main(Main.scala:1)
```

**Comment by shannah** [comment]
May not be issue with scala in general, but rather the use of reflection.

I'm not a scala guy, but I'd look at this code and see if theres a way to avoid the reflection.

This might be helpful.
https://stackoverflow.com/questions/50138533/how-does-one-completely-avoid-runtime-reflection-in-scala

**Comment by konsoletyper** [comment]
@qwqawawow which TeaVM version did you try? Can you attach a self-contained example?

**Comment by qwqawawow** [comment]
> @qwqawawow which TeaVM version did you try? Can you attach a self-contained example?

The version is 0.7.0 from maven central
[test-scala-teavm](https://github.com/qwqawawow/test-scala-teavm)

**Comment by konsoletyper** [comment]
With latest development build (0.7.0-dev-1209) this project compiled successfully. See instructions on the readme to setup development builds.

**Comment by qwqawawow** [comment]
> With latest development build (0.7.0-dev-1209) this project compiled successfully. See instructions on the readme to setup development builds.



@konsoletyper Sorry for my mistake.
But I can't run classes.js with nodejs
As you can see

`
/root/teavm/test/scala/classes.js:6
  } function $rt_castToInterface(obj, cls) { if (obj !== null && !$rt_isInstance(obj, cls)) { $rt_throwCCE(); } return obj; } function $rt_castToClass(obj, cls) { if (obj !== null && !(obj instanceof cls)) { $rt_throwCCE(); } return obj; } $rt_globals.Array.prototype.fill = $rt_globals.Array.prototype.fill || function(value, start, end) {




                      ^

TypeError: Cannot read properties of undefined (reading 'prototype')
    at /root/teavm/test/scala/classes.js:6:259
    at Object.<anonymous> (/root/teavm/test/scala/srcorg/teavm/classlib/java/nio/charset/impl/TBufferedEncoder.java:119:1)
    at Module._compile (node:internal/modules/cjs/loader:1218:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1272:10)
    at Module.load (node:internal/modules/cjs/loader:1081:32)
    at Module._load (node:internal/modules/cjs/loader:922:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:82:12)
    at node:internal/main/run_main_module:23:47

Node.js v19.2.0
`

**Comment by konsoletyper** [comment]
TeaVM was never meant to support Node.js, since it makes no sense: on the server you can run JVM which superior to Node.js in all aspects. Any idea why one would need to compile Scala to Node.js?

**Comment by OskarDamkjaer** [comment]
@konsoletyper Here's another usecase, we're compiling part of Neo4j from java/scala to javascript, so that we can give real error messages in a vscode extension. Since it runs on users machines in vscode, we'd need to ship a whole jvm with the editor extension and spawn that as a separate thread - likely possible but much more complicated

**Comment by OskarDamkjaer** [comment]
@konsoletyper I'll try and explain it better: We're building an editor extension to VSCode to help users write Cypher queries better. VSCode as an electron app has its extensions written in javascript and executed in node. We'd like to enhance our extension with error messages that we get from a big java program (neo4j) - so we used TeaVM to compile it to javascript so we can run it in our extension. 

While I think that in principle it could be possible for our extension to ship an open source JDK bundled in the extension and then spawn an external process to run the java program outside of VSCode but we'd like to avoid it.

We also package the same language support library (with error messages, syntax highlighting, etc) for use in a language server. We're not using scala3 yet, so the node build worked well for us so far. This means users who install our language server on their computer, only need node and don't need to install a jdk as well.

I just wanted to highlight a why we're interested in running scala3 on node. Let me know if it's still hard to follow, I'd be happy to clarify further!

**Comment by konsoletyper** [comment]
@OskarDamkjaer ok, I see. This particular issue (i.e. patch for `scala.runtime.Statics`) was fixed by eab7ef0a2e7d9e3b7b0843c889b3d14ba538d3e6. Did you try TeaVM to compile your Scala code to JavaScript?

As for Node.js, you can try latest TeaVM (see [this article](https://teavm.org/docs/intro/preview-builds.html)). Latest TeaVM produces ES6 modules, but you should turn on this functionality (sorry, no documentation for preview builds). I never had a chance to test with Node.js, but don't see any reason why JS produced by latest TeaVM won't run on it. Otherwise, feel free to open corresponding issues. If you decide to start using TeaVM, feel free to open discussions or ask me in chat. But I should warn you, that using TeaVM to compile neo4j might be tough.

**Comment by OskarDamkjaer** [comment]
@konsoletyper I've not tried it with scala3 yet, my main motivation was giving some context as to why anyone would want to run TeaVM in node :) 

Yes all of Neo4j I think would be challenging but we're fiddling a small-ish parts of it currently. Happy to hear about the ES6 modules, I'll be sure to check out the preview and open issues/discussion if I run into any issues or come up with any ideas
